service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/UploadAdditionalApplicationsController.java

```diff
@@ -0,0 +1,179 @@
+        CaseData caseData = getCaseData(caseDetails);
+
+        caseDetails.getData().putAll(applicationsFeeCalculator.calculateFee(caseData));
+
+        return respond(caseDetails);
+    }
+
+    @PostMapping("/validate/mid-event")
+    public AboutToStartOrSubmitCallbackResponse handleValidateMidEvent(@RequestBody CallbackRequest callbackRequest) {
+        CaseDetails caseDetails = callbackRequest.getCaseDetails();
+        CaseData caseData = getCaseData(caseDetails);
+
+        List<String> errors = new ArrayList<>();
+
+        if (caseData.getTemporaryPbaPayment() != null) {
+            var updatedPbaNumber = pbaNumberService.update(caseData.getTemporaryPbaPayment().getPbaNumber());
+            caseDetails.getData().put("temporaryPbaPayment",
+                caseData.getTemporaryPbaPayment().toBuilder().pbaNumber(updatedPbaNumber).build());
+            errors.addAll(pbaNumberService.validate(updatedPbaNumber));
+        }
+
+        return respond(caseDetails, errors);
+    }
+
+    @PostMapping("/about-to-submit")
+    public AboutToStartOrSubmitCallbackResponse handleAboutToSubmit(
+        @RequestBody CallbackRequest callbackRequest) {
+        CaseDetails caseDetails = callbackRequest.getCaseDetails();
+        CaseData caseData = getCaseData(caseDetails);
+
+        List<Element<AdditionalApplicationsBundle>> additionalApplications = defaultIfNull(
+            caseData.getAdditionalApplicationsBundle(), new ArrayList<>()
+        );
+
+        C2DocumentBundle c2DocumentBundle = null;
+        OtherApplicationsBundle otherApplicationsBundle = null;
+
+        if (hasC2Order(caseData)) {
+            c2DocumentBundle = uploadC2DocumentsService.buildC2DocumentBundle(caseData);
+        }
+
+        if (hasOtherApplicationsOrder(caseData)) {
+            otherApplicationsBundle = uploadC2DocumentsService.buildOtherApplicationsBundle(caseData);
+        }
+
+        additionalApplications.add(0, element(uploadC2DocumentsService.buildAdditionalApplicationsBundle(
+            caseData, c2DocumentBundle, otherApplicationsBundle)));
+
+        caseDetails.getData().put("additionalApplicationsBundle", additionalApplications);
+
+        removeTemporaryFields(caseDetails, TEMPORARY_C2_DOCUMENT,
+            "c2Type",
+            "additionalApplicationType",
+            "usePbaPayment",
+            AMOUNT_TO_PAY,
+            "pbaNumber",
+            "clientCode",
+            "fileReference",
+            "temporaryPbaPayment",
+            TEMPORARY_OTHER_APPLICATIONS_BUNDLE);
+
+        return respond(caseDetails);
+    }
+
+    @PostMapping("/submitted")
+    public void handleSubmittedEvent(
+        @RequestBody CallbackRequest callbackRequest) {
+        CaseDetails caseDetails = callbackRequest.getCaseDetails();
+        CaseData caseData = getCaseData(caseDetails);
+
+        final PBAPayment pbaPayment = caseData.getAdditionalApplicationsBundle().get(0).getValue().getPbaPayment();
```

Comment:
caseData.getAdditionalApplicationsBundle().get(0) is repeated twice, can be

final AdditionalApplicationsBundle lastBundle = caseData.getAdditionalApplicationsBundle().get(0).getValue();